central/telemetry/phonehome/gatherer.go

```diff
@@ -0,0 +1,158 @@
+	groupDataStore "github.com/stackrox/rox/central/group/datastore"
+	roles "github.com/stackrox/rox/central/role/datastore"
+	si "github.com/stackrox/rox/central/signatureintegration/datastore"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+	pkgPH "github.com/stackrox/rox/pkg/telemetry/phonehome"
+)
+
+var (
+	log              = logging.LoggerForModule()
+	gathererInstance *gatherer
+	onceGatherer     sync.Once
+)
+
+const period = 5 * time.Minute
+
+type gatherer struct {
+	telemeter pkgPH.Telemeter
+	period    time.Duration
+	ticker    *time.Ticker
+	stopSig   concurrency.Signal
+	ctx       context.Context
+	cancel    context.CancelFunc
+	mu        sync.Mutex
+	f         func(*gatherer)
+}
+
+// Gatherer interface for interacting with telemetry gatherer.
+type Gatherer interface {
+	Start()
+	Stop()
+}
+
+func (g *gatherer) reset() {
+	ctx, cancel := context.WithCancel(context.Background())
+	g.ctx = sac.WithAllAccess(ctx)
+	g.cancel = cancel
+	g.stopSig = concurrency.NewSignal()
+}
+
+func newGatherer(t pkgPH.Telemeter, p time.Duration, f func(*gatherer)) *gatherer {
+	g := &gatherer{
+		telemeter: t,
+		period:    p,
+		f:         f,
+	}
+	g.reset()
+	return g
+}
+
+// GathererSingleton returns the telemetry gatherer instance.
+func GathererSingleton() Gatherer {
+	if Enabled() {
+		onceGatherer.Do(func() {
+			gathererInstance = newGatherer(TelemeterSingleton(), period, func(g *gatherer) { g.gather() })
+		})
+	}
+	return gathererInstance
+}
+
+func (g *gatherer) loop() {
+	g.ticker = time.NewTicker(g.period)
+	for !g.stopSig.IsDone() {
+		select {
+		case <-g.ticker.C:
+			go g.f(g)
+		case <-g.stopSig.Done():
+			g.ticker.Stop()
+			g.ticker = nil
+			g.cancel()
+			return
+		}
+	}
+	log.Debug("Loop stopped.")
+}
+
+func (g *gatherer) Start() {
+	if g == nil {
+		return
+	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	// Ignore Start if the ticker is active.
+	if g.ticker == nil {
+		g.reset()
+		go g.loop()
+		log.Debug("Telemetry data collection ticker enabled.")
+	}
+}
+
+func (g *gatherer) Stop() {
+	if g == nil {
+		return
+	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	g.stopSig.Signal()
+}
+
+func addTotal[T any](ctx context.Context, props map[string]any, key string, f func(context.Context) ([]*T, error)) {
+	if ps, err := f(ctx); err != nil {
+		log.Errorf("Failed to get %s: %v", key, err)
+	} else {
+		props["Total "+key] = len(ps)
+	}
+}
+
+func (g *gatherer) gather() {
+	log.Debug("Starting telemetry data collection.")
+	defer log.Debug("Done with telemetry data collection.")
+
+	totals := make(map[string]any)
+	rs := roles.Singleton()
+
+	addTotal(g.ctx, totals, "PermissionSets", rs.GetAllPermissionSets)
+	addTotal(g.ctx, totals, "Roles", rs.GetAllRoles)
+	addTotal(g.ctx, totals, "Access Scopes", rs.GetAllAccessScopes)
+	addTotal(g.ctx, totals, "Signature Integrations", si.Singleton().GetAllSignatureIntegrations)
```

Comment:
I think it would be more beneficial to have the _things we want to collect_ more or less injected within the gatherer in form of a separate abstraction, such as:
```
type gathererFunc func(ctx context.Context) (map[string]any, error)
```

Right now, if we were to arbitrarily add additional datastores we want to collect from to our targets, we would always have to touch this specific `gather()` function. I'm not sure whether it's planned in the future to do this more or less arbitrarily (e.g. defining _what_ to collect based on each version via e.g. env variables / other configuration (so that we can differentiate between information collected from on-prem vs ACSCS), but this would certainly feel like a better abstraction.

Additionally, the currently custom login w.r.t auth providers and groups should be moved to either a separate function implementing the abstraction, or even the datastore itself. I would have expected the gatherer to be just a "stupid" component without knowing specific details about the types of proto messages it is gathering (especially because an abstraction like this [already existed for the "old" telemetry](https://github.com/stackrox/stackrox/tree/master/central/telemetry/gatherers))

Reply:
Please check the updated version. I had to shuffle the code a little to avoid circular dependencies with authn package, which imports roles.